File: ios/legacy/HybridViewModelListProperty.swift
import NitroModules
import RiveRuntime

class HybridViewModelListProperty: HybridViewModelListPropertySpec, ValuedPropertyProtocol {
  var property: ListPropertyType!
  lazy var helper = PropertyListenerHelper(property: property!)

  init(property: ListPropertyType) {
    self.property = property
    super.init()
  }

  var length: Double {
    MainThread.run { Double(property.count) }
  }

  func getInstanceAt(index: Double) throws -> (any HybridViewModelInstanceSpec)? {
    MainThread.run { () -> (any HybridViewModelInstanceSpec)? in
      guard let instance = property.instance(at: Int32(index)) else { return nil }
      return HybridViewModelInstance(viewModelInstance: instance)
    }
  }

  private func requireViewModelInstance(_ instance: any HybridViewModelInstanceSpec) throws -> RiveDataBindingViewModel.Instance {
    guard let hybridInstance = instance as? HybridViewModelInstance,
          let viewModelInstance = hybridInstance.viewModelInstance else {
      throw NSError(domain: "HybridViewModelListProperty", code: 1,
                    userInfo: [NSLocalizedDescriptionKey: "Expected HybridViewModelInstance with valid viewModelInstance"])
    }
    return viewModelInstance
  }

  func addInstance(instance: any HybridViewModelInstanceSpec) throws {
    let viewModelInstance = try requireViewModelInstance(instance)
    MainThread.run { property.append(viewModelInstance) }
  }

  func addInstanceAt(instance: any HybridViewModelInstanceSpec, index: Double) throws -> Bool {
    let viewModelInstance = try requireViewModelInstance(instance)
    return MainThread.run { property.insert(viewModelInstance, at: Int32(index)) }
  }

  func removeInstance(instance: any HybridViewModelInstanceSpec) throws {
    let viewModelInstance = try requireViewModelInstance(instance)
    MainThread.run { property.remove(viewModelInstance) }
  }

  func removeInstanceAt(index: Double) throws {
    MainThread.run { property.remove(at: Int32(index)) }
  }

  func swap(index1: Double, index2: Double) throws -> Bool {
    MainThread.run {
      let idx1 = UInt32(index1)
      let idx2 = UInt32(index2)
      guard idx1 < property.count && idx2 < property.count else {
        return false
      }
      property.swap(at: idx1, with: idx2)
      return true
    }
  }

  func getLengthAsync() throws -> Promise<Double> {
    return Promise.onMain { self.length }
  }

  func getInstanceAtAsync(index: Double) throws -> Promise<(any HybridViewModelInstanceSpec)?> {
    return Promise.onMain { try self.getInstanceAt(index: index) }
  }

  func addInstanceAsync(instance: any HybridViewModelInstanceSpec) throws -> Promise<Void> {
    return Promise.onMain { try self.addInstance(instance: instance) }
  }

  func addInstanceAtAsync(instance: any HybridViewModelInstanceSpec, index: Double) throws -> Promise<Void> {
    return Promise.onMain { let _ = try self.addInstanceAt(instance: instance, index: index) }
  }

  func removeInstanceAsync(instance: any HybridViewModelInstanceSpec) throws -> Promise<Void> {
    return Promise.onMain { try self.removeInstance(instance: instance) }
  }

  func removeInstanceAtAsync(index: Double) throws -> Promise<Void> {
    return Promise.onMain { try self.removeInstanceAt(index: index) }
  }

  func swapAsync(index1: Double, index2: Double) throws -> Promise<Void> {
    return Promise.onMain { let _ = try self.swap(index1: index1, index2: index2) }
  }

  func addListener(onChanged: @escaping () -> Void) throws -> () -> Void {
    helper.addListener({ _ in onChanged() })
  }
}

File: android/src/legacy/java/com/margelo/nitro/rive/HybridViewModelListProperty.kt
package com.margelo.nitro.rive

import androidx.annotation.Keep
import app.rive.runtime.kotlin.core.ViewModelListProperty
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.flow.map

@Keep
@DoNotStrip
class HybridViewModelListProperty(private val listProperty: ViewModelListProperty) :
  HybridViewModelListPropertySpec(),
  BaseHybridViewModelProperty<Unit> by BaseHybridViewModelPropertyImpl() {
  // Deprecated: Use getLengthAsync instead
  override val length: Double
    get() = listProperty.size.toDouble()

  override fun getLengthAsync(): Promise<Double> {
    return Promise.async { listProperty.size.toDouble() }
  }

  private fun requireHybridInstance(instance: HybridViewModelInstanceSpec): HybridViewModelInstance {
    return instance as? HybridViewModelInstance
      ?: throw IllegalArgumentException("Expected HybridViewModelInstance but got ${instance::class.simpleName}")
  }

  // Deprecated: Use getInstanceAtAsync instead
  override fun getInstanceAt(index: Double): HybridViewModelInstanceSpec? {
    val idx = index.toInt()
    if (idx < 0 || idx >= listProperty.size) return null
    return HybridViewModelInstance(listProperty.elementAt(idx))
  }

  override fun getInstanceAtAsync(index: Double): Promise<HybridViewModelInstanceSpec?> {
    return Promise.async {
      val idx = index.toInt()
      if (idx < 0 || idx >= listProperty.size) null
      else HybridViewModelInstance(listProperty.elementAt(idx))
    }
  }

  override fun addInstance(instance: HybridViewModelInstanceSpec) {
    val hybridInstance = requireHybridInstance(instance)
    listProperty.add(hybridInstance.viewModelInstance)
  }

  override fun addInstanceAt(instance: HybridViewModelInstanceSpec, index: Double): Boolean {
    val hybridInstance = requireHybridInstance(instance)
    val idx = index.toInt()
    if (idx < 0 || idx > listProperty.size) return false
    listProperty.add(idx, hybridInstance.viewModelInstance)
    return true
  }

  override fun removeInstance(instance: HybridViewModelInstanceSpec) {
    val hybridInstance = requireHybridInstance(instance)
    listProperty.remove(hybridInstance.viewModelInstance)
  }

  override fun removeInstanceAt(index: Double) {
    listProperty.removeAt(index.toInt())
  }

  override fun swap(index1: Double, index2: Double): Boolean {
    val idx1 = index1.toInt()
    val idx2 = index2.toInt()
    if (idx1 < 0 || idx1 >= listProperty.size || idx2 < 0 || idx2 >= listProperty.size) {
      return false
    }
    listProperty.swap(idx1, idx2)
    return true
  }

  override fun getLengthAsync(): Promise<Double> {
    return Promise.async { length }
  }

  override fun getInstanceAtAsync(index: Double): Promise<HybridViewModelInstanceSpec?> {
    return Promise.async { getInstanceAt(index) }
  }

  override fun addInstanceAsync(instance: HybridViewModelInstanceSpec): Promise<Unit> {
    return Promise.async { addInstance(instance) }
  }

  override fun addInstanceAtAsync(instance: HybridViewModelInstanceSpec, index: Double): Promise<Unit> {
    return Promise.async { addInstanceAt(instance, index) }
  }

  override fun removeInstanceAsync(instance: HybridViewModelInstanceSpec): Promise<Unit> {
    return Promise.async { removeInstance(instance) }
  }

  override fun removeInstanceAtAsync(index: Double): Promise<Unit> {
    return Promise.async { removeInstanceAt(index) }
  }

  override fun swapAsync(index1: Double, index2: Double): Promise<Unit> {
    return Promise.async { swap(index1, index2) }
  }

  override fun addListener(onChanged: () -> Unit): () -> Unit {
    val remover = addListenerInternal { _ -> onChanged() }
    ensureValueListenerJob(listProperty.valueFlow.map { })
    return remover
  }
}
